### ISSUE TITLE
GameState-sensitive services

### ISSUE THREAD

**Issue by Aaron1011** [SpongePowered/SpongeAPI#1036]
The current Service API and implementation assume that service providers may be accessed at any time. However, as [this issue](https://github.com/SpongePowered/SpongeCommon/issues/291) shows, this doesn't always work out. Particule service providers may require the server to be running, and will not work properly otherwise.

I've outlined some possible solutions below. Each one has some drawbacks, and I'n not really sure what the best one would be.

* **Require services to check the current game state before performing operatings**
This would require `UserStorage` to only use the server between `SERVER_STARTED` and `SERVER_STOPPED`. While the service itself would always be available, certain methods might fail or throw an exception, depending in which game state they were used

  Advantages:
  * The existing service API remains unchanged
  * Plugins are free to store a reference to a service provider, and need only update it on a `ChangeServiceProviderEvent`
  * Services can still provide limited functionality to plugins, even if using their full capabilities are not available.

  Disadvantages:
  * Not all service providers can meaningfully exist at all times. For example, a service dealing only with worlds could not function at all if the server is not running.
 * Services have no guarantee about the game state they are running in, and must attempt to fail gracefully if a particular method cannot work in the current environment.

* **Allow services providers to be removed after one is provided**
  At the moment, plugin can safely call `ServiceManager#provideUnchecked` any time after a provider is first registered for a given service. With this solution, a service could be "un-provided" at some point after a provider is registered, causing the `ServiceManager` to act as though no provider has ever been registered.

  Advantages:
  * Plugins providing a service can choose to make it unavailable, depending on the current game state and environment. The previously mentioned world-related service could only be made available while the server is running.
  * The client environment is properly handled, since a provider can be "un-provided" each time a particular change in the environment occurs.

  Disadvantages:
 * Plugins can **never** store a reference to a service provider. Since the entire point of this solution is that a provider may become unavailable, any stored reference to a service provider will become invalid if the provider is removed. This would require plugins to attempt to retrieve a service from the `ServiceManager` each time they wished to use it.
 * The `ServiceManager` and `ChangeServiceProviderEvent` must be changed to accommodate the fact that a provider can be replaced by no provider.
* The unavailability of a provider for a given service conveys no additional information about the reason why it is not available.

Pinging @SpongePowered/developers for feedback,

**Comment by DarkArc** [comment]
I like option 2 better: Allow services providers to be removed after one is provided.

I don't think we should be supporting the storage of references to services as is. This poses problems not only for provided services being removed, but for service providers being changed, as the referred service, might not actually be the currently provided service. Thus, the developer must listen for any changes to the service provider, and we've created a new area for human error.

**Comment by zml2008** [comment]
Yeah, I like option 2 as well. There isn't a huge amount of additional handling required now that we already support replaceable service handlers (though it might make sense to specify whether or not a certain service type is nullable).

**Comment by ST-DDT** [comment]
We could avoid most issues with outdated references if we don't expose the actual service impl but use a proxy impl. Of course they will still throw exceptions if no service is registered at invocation time. And this will avoid the work to update service references for services that must be present (PermissionService). Maybe we could even add a new interface to allow such isdirty /isregistered checks for services.

**Comment by ryantheleach** [comment]
If you are worried about people storing references, Can't we give people a Supplier<Service> instead?

IMO services that have been registered shouldn't be unavailable. If they get de-registered a new service should be ready to take it's place immediately, It makes it far easier to program to if you can assume that once a service is available either it, or another is going to remain available.

(It would be even better if there was an event or something that means plugins need to re-initialize the service when replacements happened. That way I can be sure my PermissionDescriptions have been registered even if the Permissions service swaps out)

**Comment by Aaron1011** [comment]
@ryantheleach: There's `ChangeServiceProviderEvent`, which does what you described. However, as you pointed out, there's still the problem of needing to disable some of your plugin's functionality each time the provider is removed.

**Comment by Zidane** [comment]
@ryantheleach 

I disagree on ```IMO services that have been registered shouldn't be unavailable.```. If, for some reason, an Econ plugins removes its registered service then there is no fallback and therefore Sponge should report that there is no service should a plugin query for the EconService.

There really is a simple fix here; plugins should simply query for that service when they are accessing it lol.

**Comment by ryantheleach** [comment]
I think we need to bring this back to why and in which circumstances a service will/could/should be removed without a replacement.

The _only_ one I can think of is universe changing on the client. But even then restarting the game is usually required for significant changes of what mods are enabled.

**Comment by hsyyid** [comment]
IMO, there needs to be a way to either get the `pluginId` of the plugin providing the service, or a way to get the PluginInstance itself.
